tuta-sdk/rust/src/login/login_facade.rs

```diff
@@ -0,0 +1,320 @@
+use std::cmp::PartialEq;
+use std::sync::Arc;
+
+use base64::{DecodeError, Engine};
+use base64::prelude::{BASE64_STANDARD, BASE64_URL_SAFE_NO_PAD};
+use thiserror::Error;
+
+use crate::{ApiCallError, IdTuple};
+use crate::ApiCallError::InternalSdkError;
+use crate::crypto::{Aes256Key, generate_key_from_passphrase, sha256};
+use crate::crypto::key::GenericAesKey;
+use crate::entities::sys::{Session, User};
+#[mockall_double::double]
+use crate::typed_entity_client::TypedEntityClient;
+use crate::typed_entity_client::TypedEntityClientHandlers;
+use crate::generated_id::GeneratedId;
+use crate::login::credentials::Credentials;
+use crate::login::login_controller::{ExternalUserKeyDeriver, KdfType};
+use crate::user_facade::UserFacade;
+use crate::util::{array_cast_slice, BASE64_EXT};
+
+/// Error that may occur during login and session creation
+#[derive(Error, Debug, uniffi::Error)]
+pub enum LoginError {
+    #[error("SessionIdError: {error_message}")]
+    SessionIdError {
+        error_message: String,
+    },
+    #[error("PassphraseError: {error_message}")]
+    PassphraseError {
+        error_message: String,
+    },
+    #[error("KeyError: {error_message}")]
+    KeyError {
+        error_message: String,
+    },
+    #[error("ApiCallError: {source}")]
+    ApiCallError {
+        #[from]
+        source: ApiCallError,
+    },
+}
+
+struct LoginFacade {
+    entity_client: Arc<TypedEntityClient>,
+}
+
+pub enum SessionType {
+    Login,
+    Temporary,
+    Persistent,
+}
+
+#[derive(PartialEq)]
+pub enum AccountType {
+    SYSTEM,
+    FREE,
+    STARTER,
+    PAID,
+    EXTERNAL,
+}
+
+impl AccountType {
+    pub fn from(value: i64) -> AccountType {
+        match value {
+            0 => AccountType::SYSTEM,
+            1 => AccountType::FREE,
+            2 => AccountType::STARTER,
+            3 => AccountType::PAID,
+            4 => AccountType::EXTERNAL,
+            _ => panic!("Invalid account type")
+        }
+    }
+}
+
+pub const GENERATED_ID_BYTES_LENGTH: usize = 9;
+
+impl LoginFacade {
+    pub fn new(
+        entity_client: Arc<TypedEntityClient>,
+    ) -> Self {
+        LoginFacade {
+            entity_client,
+        }
+    }
+
+    /// Generate session id tuple from access token
+    fn get_session_id(&self, access_token: &str) -> Result<IdTuple, DecodeError> {
```

Comment:
we can probably move this outside of the impl, it doesn't use `self` and we can rename it into something like `parse_session_id`

Reply:
Not sure where best to put it, I just put it in the same file for now.